Issue #9280: New property 'accessModifiers' as substitution of 'scope' and 'excludeScope' in JavadocVariableCheck

kkoutsilis opened the pull request:
> [!CAUTION]
> Breaking change

Aims to close #9280 
This PR removes `scope` and `excludeScope` from [JavadocVariableCheck](https://checkstyle.org/checks/javadoc/javadocvariable.html#JavadocVariable) while substitutes with new `accessModifers` property.

kkoutsilis commented:
For example in [InputJavadocVariableNoJavadoc2](https://github.com/checkstyle/checkstyle/blob/8e5396f11e39711d16420cbb704c5b2da232621c/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableNoJavadoc2.java#L13) line 72, surrounding scope is "package" while surrounding access modifier is "public" 
<img width="783" alt="Screenshot 2024-12-24 at 18 52 58" src="https://github.com/user-attachments/assets/5068f4a8-f3dd-4f54-9e88-cb57578ebd94" />

kkoutsilis commented:
I believe the difference is coming from this extra check https://github.com/checkstyle/checkstyle/blob/8e5396f11e39711d16420cbb704c5b2da232621c/src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java#L373

rdiachenko commented:
@kkoutsilis I might not have the full context, but what's the reason for using surrounding access modifiers in the first place? Why can't we use direct access modifiers only? That way the logic and usage would be more straightforward for users.

kkoutsilis commented:
> @kkoutsilis I might not have the full context, but what's the reason for using surrounding access modifiers in the first place? Why can't we use direct access modifiers only? That way the logic and usage would be more straightforward for users.

Hey @rdiachenko, thanks for your feedback. I am mainly trying to keep the same logic of how scope worked before, so since scope used surrounding Scope, I am trying to replicate that with access modifiers to keep it as consistent as possible. I am not sure what was the decision of using surrounding Scope, I can only make guesses. 
Your feedback is valid, I am not sure if I can take this kind of "design" decisions as I am fairly new contributor to the project but I am happy to go this down this path of not using surrounding access modifiers if you agree with it :) 
If not, we can go with the initial implementation for the abandoned PR that had a [method](https://github.com/checkstyle/checkstyle/pull/9277/files#diff-1bda71a340401c2d7c43026b4306d4b9bbc84515fa08c78d5e70e552cab47025R280) to map scope to access modifier, and this works fine. 

It stills bugs me that `getSurroundingScope` and and `getSurroundingAccessModifier` are inconsistent between them, and can't tell if that was intentional or a bug. Do you have any insights on that by any chance? Thank you in advance!

rdiachenko commented:
> Hey @rdiachenko, thanks for your feedback. I am mainly trying to keep the same logic of how scope worked before, so since scope used surrounding Scope, I am trying to replicate that with access modifiers to keep it as consistent as possible. I am not sure what was the decision of using surrounding Scope, I can only make guesses. Your feedback is valid, I am not sure if I can take this kind of "design" decisions as I am fairly new contributor to the project but I am happy to go this down this path of not using surrounding access modifiers if you agree with it :) If not, we can go with the initial implementation for the abandoned PR that had a [method](https://github.com/checkstyle/checkstyle/pull/9277/files#diff-1bda71a340401c2d7c43026b4306d4b9bbc84515fa08c78d5e70e552cab47025R280) to map scope to access modifier, and this works fine.
> 
> It stills bugs me that `getSurroundingScope` and and `getSurroundingAccessModifier` are inconsistent between them, and can't tell if that was intentional or a bug. Do you have any insights on that by any chance? Thank you in advance!

Scope and access modifier are two different concepts as for me. I don't even understand why we have `getSurroundingAccessModifier`. Let's make it simple for this check, we break backward compatibility anyway. If we need advanced filtering based on outer access modifier, let's do it as a separate step, when we have real use cases. As for now, I don't think of a use case where we would need the outer (surrounding modifier) context.

@romani do you have any input on this?

romani commented:
>but it seems there are use cases that these methods return different values (matching scope to access modifier)

As in any other source code base, until methods are not ideal and just serve some goal, that can be viewed by usage of methods. 
If for any reason such methods don't work for this Check, better to make new methods that work for this Check. As completely separate activity we can refactor other Checks to use new methods if it makes sense.

>surrounding scope is "package" while surrounding access modifier is "public"

This is long and heavy problem to define scope/visibility based on all upper classes.
I think we should look at class/type's fields  own visibility, completely ignoring all upper types visibility/modifiers.

To make it even more complicated, modern java has a lot not explicit visibilities, we have separate Check on this, and it is very complicated by its own.

-------

Let's not consider any non target visibility. On edge cases, users might use suppressions.

romani commented:
@kkoutsilis, let me know if you need any assistance

romani commented:
please rebase, I moved xdoc folder location.

kkoutsilis commented:
Hey @romani, I am away atm, will do as soon as I am back

kkoutsilis commented:
Hey @romani, I rebased. Will try to look this week into why the CI is failing but I aslo believe there are some requirements from the issue than I need to satisfy. Hopefully I will have some time to look into this.

romani commented:
All CI failures are due to breaking changes.
So we clearly see who is affected.
As we get closer to merge 
We need to send them PR with migration, so it will confirm that migration is possible without bad side effects.
See you soon, we appreciate your help.

kkoutsilis commented:
Hey @romani, I believe the PR can be reviewed

rdiachenko commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -84,35 +81,28 @@ public class JavadocVariableCheck
-        this.excludeScope = excludeScope;
+    public void setAccessModifiers(AccessModifierOption... accessModifiers) {
+        this.accessModifiers =
+            Arrays.copyOf(accessModifiers, accessModifiers.length);
```
Do we actually need to make a copy here?

kkoutsilis replied:
I made it this way to keep it consistent with the codebase https://github.com/checkstyle/checkstyle/blob/62bed57795d1c439aa5ad5adb8ca5c3d671bd62f/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/ParameterNameCheck.java#L120-L123  https://github.com/checkstyle/checkstyle/blob/62bed57795d1c439aa5ad5adb8ca5c3d671bd62f/src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java#L114-L117
I guess it is done so it does not mess up accidentally with the passed down array. 
What would be the benefit of not copying ?

rdiachenko replied:
> I made it this way to keep it consistent with the codebase

I see only two checks that use such approach

> What would be the benefit of not copying ?

If we don't copy it, we don't need an adhock unit test that doesn't bring much value. Additionally, I don't think the passed down array is used anywhere else outside this check, so such protection looks redundant to me.

kkoutsilis replied:
Seems that there are checks in place to avoid this behaviour
```
[WARNING] PMD Failure: com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocVariableCheck:104 Rule:ArrayIsStoredDirectly Priority:3 The user-supplied array 'accessModifiers' is stored directly..
```
With a quick [SO search](https://stackoverflow.com/a/11581068), the solution is copying the array. 
Not sure if there is any other solution in order to avoid that extra test.

romani replied:
As class receiving such array from outside, for crazy security rules it should be copied, as all  around are enemies.

rdiachenko replied:
Ok, thanks for clarifying this @kkoutsilis. Does it make sense to use `copyOfArray` from `UnmodifiableCollectionUtil`? Will it help to avoid the unit test?

kkoutsilis replied:
Looking at this note on the class https://github.com/checkstyle/checkstyle/blob/dae549f530608c833407925d4caac75a8624f9e3/src/main/java/com/puppycrawl/tools/checkstyle/utils/UnmodifiableCollectionUtil.java#L30-L36
 I guess it does, I will give it a go. Thanks for the pointer @rdiachenko

kkoutsilis replied:
Updated

rdiachenko commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheckTest.java:
```diff
@@ -338,13 +339,47 @@ public void testLambdaLocalVariablesDoNotNeedJavadoc() throws Exception {
+     * @throws Exception if an error occurs.
+     */
+    @Test
+    public void testCloneInAccessModifiersProperty() throws Exception {
```
Do we need this test if we don't copy the array?

kkoutsilis replied:
I don't believe we do

kkoutsilis replied:
Removed

kkoutsilis commented:
Hey @romani , while this PR is in review, do you want me to look into another issue?

rdiachenko reviewed (approved):
lgtm

romani commented:
@kkoutsilis , please start creating PRs to fix/migrate configurations in project that are in CI.

Please update respectful CI jobs to check out your fork of such repositories and checkout to your branch.

Ideal scenario: we just update config in their repositories. If any other violations happens we need to discuss.

Choose any repository/project you like and lets try to migrate it first.
Pin me for any help you need.

kkoutsilis commented:
> @kkoutsilis , please start creating PRs to fix/migrate configurations in project that are in CI.
> 
> Please update respectful CI jobs to check out your fork of such repositories and checkout to your branch.
> 
> Ideal scenario: we just update config in their repositories. If any other violations happens we need to discuss.
> 
> Choose any repository/project you like and lets try to migrate it first. Pin me for any help you need.

Hey @romani, just to make sure I inderstand correctly, we need to:
1. Fork the projects that are failing in CI and update the configuration. for example [spring-cloud-gcp](https://github.com/GoogleCloudPlatform/spring-cloud-gcp)
2. Update CI to checkout to fork/bracnh with updated config
3. Make sure that CI pass here

Is that correct or did I misunderstand ?

romani commented:
yes!
if you see that you are going to make a lot of work, pin me to review scope of updates, we shoould not be involved in something big, it might be sign that we are doing smething wrong.

kkoutsilis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -84,35 +82,28 @@ public class JavadocVariableCheck
-     * @param scope a scope.
-     * @since 3.0
+     * @param accessModifiers access modifiers of methods which should be checked.
+     * @since 11.0.0
```
@romani Just to doublechekc here as well, which is the target release version of this change? Added 11.0.0 as its a breaking change. What is the plan with releases?

romani replied:
lets put here next release with major changes, 10.22.0 .
your changes are breaking changes, so after merge cI will be red to demand version bump to 10.22.0

kkoutsilis replied:
Updated

kkoutsilis commented:
Hey @romani, fixed [no-error-pmd](https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30093/workflows/38ceb275-b0b9-40a2-a1c2-57cb85b02e71/jobs/780545) check. [Chenge](https://github.com/pmd/build-tools/compare/main...kkoutsilis:build-tools:update-checkstyle-javadocvariable) in config was replacing `scope` with `accessModifiers` in build-tools and make sure that pmd [points](https://github.com/pmd/pmd/compare/main...kkoutsilis:pmd:update-checkstyle-javadocvariable) to the new config, so for this case it was the idieal scenario. Will probably move to [no-error-xwiki](https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30093/workflows/38ceb275-b0b9-40a2-a1c2-57cb85b02e71/jobs/780548) next.

kkoutsilis commented:
Hey @romani, fixed  [no-error-xwiki](https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30111/workflows/e57246c6-0e6f-40ac-8c83-a8f8277edce1/jobs/781300) as well. The story is a bit different here. Since we changed the behaviour of the check to not account for the surrounding scope, some javadocs or suppressions must be added. Here are the changes for [xwiki-platform](https://github.com/xwiki/xwiki-platform/compare/master...kkoutsilis:xwiki-platform:update-checkstyle-javadocvariable) and for [xwiki-commons](https://github.com/xwiki/xwiki-commons/compare/master...kkoutsilis:xwiki-commons:update-checkstyle-javadocvariable)

romani commented:
> spring-cloud-gcp

migrated to google style config.
https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/869/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R264
https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/868
very old version of checkstyle 9.3: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/751c57c209db740c80171d0a3e60eb9a33f1374a/spring-cloud-gcp-samples/pom.xml#L25

we can not support that old version.


We need to drop spring-cloud-gcp from our CI, it is too heavy to migrate them to most recent version.

@elefeint , fyi, but let us know when you be able to catch up to recent version, so we can  restore it in our CI.

romani commented:
@kkoutsilis , please send PRs to xwiki and pmd.
rebase on latest,

to avoid whilelist polution that is requested at https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=26005&view=logs&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358&l=55

please use this trick:
` git checkout "66d""ed33c74662cb3da612f3d34a5ae""fa""a629b443"`
your user name can be added to whiletilst file.

elefeint commented:
@meltsufin FYI, checkstyle needs to be upgraded in spring-cloud-gcp

kkoutsilis commented:
@romani can you re-run [no-error-xwiki](https://app.circleci.com/pipelines/github/checkstyle/checkstyle/30340/workflows/7f8a0abf-e90c-4a7d-8c44-b4cf1c77f764/jobs/790912) please so I don't re-run the entire Ci

romani commented on .ci/validation.sh:
```diff
@@ -758,7 +767,6 @@ no-error-spring-cloud-gcp)
   echo "Checkout target sources ..."
   checkout_from https://github.com/googlecloudplatform/spring-cloud-gcp
   cd .ci-temp/spring-cloud-gcp
-  git checkout "7c99f37087ac8f""eb""db""f7e185375a217b744895a4"
```
please removev whole  `no-error-spring-cloud-gcp)`
and remove https://github.com/checkstyle/checkstyle/blob/9b463180f0b96a7317f92484770f186cada3d910/.semaphore/semaphore.yml#L92

we keep in our CI only projects what allows us to overide version and has no violations from CS on each commit.

kkoutsilis replied:
[Done](https://github.com/checkstyle/checkstyle/compare/b8b8813b358eeac778e4ba7a6ba7918d859963aa..7c776e0a49a6a668088b9bbcecf74742adb99128)

romani reviewed (changes requested):
item:

romani commented:
@kkoutsilis , please send PRs to xwiki and pmd to get user feedback on update.

kkoutsilis commented:
Hey @romani, I've already sent the prs

romani commented:
please share links here to let subscribe to them.

kkoutsilis commented:
- [build-tools PR](https://github.com/pmd/build-tools/pull/55)
- [xwiki-commons PR](https://github.com/xwiki/xwiki-commons/pull/1241)
- [xwiki-platforms PR](https://github.com/xwiki/xwiki-platform/pull/3907)

romani commented:
Please apply https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=26215&view=logs&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358&l=46

romani commented:
@kkoutsilis , please confirm by cli or Input that your fix covers https://github.com/checkstyle/checkstyle/issues/3183
I found it in issue description.

romani reviewed (approved):
thanks a lot !!!!

romani commented:
@kkoutsilis , please glance at xwiki failure in CI

kkoutsilis commented:
> @kkoutsilis , please glance at xwiki failure in CI

looking at it rn, its because of recent changes at master, ci was pointing to another fork with new changes, fix on the way

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -84,35 +82,28 @@ public class JavadocVariableCheck
 
     /**
-     * Setter to specify the visibility scope where Javadoc comments are checked.
+     * Setter to access modifiers of methods where parameters are checked.
```
Why we do we keep mentioning `methods`? This makes no sense. 
```suggestion
     * Setter specify the access modifiers of fields to check.
```

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -84,35 +82,28 @@ public class JavadocVariableCheck
      *
-     * @param scope a scope.
-     * @since 3.0
+     * @param accessModifiers access modifiers of methods which should be checked.
```
Same as above

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -176,6 +167,17 @@ private boolean isIgnored(DetailAST ast) {
+     * @param accessModifier the access modifier of the method.
+     * @return whether the method matches the expected access modifier.
+     */
+    private boolean matchAccessModifiers(final AccessModifierOption accessModifier) {
```
```suggestion
    private boolean matchAccessModifiers(AccessModifierOption accessModifier) {
```

Please remove `final` from the parameter declaration, this is an outdated practice.

kkoutsilis replied:
removed

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocVariableCheck.java:
```diff
@@ -185,14 +187,15 @@ private boolean isIgnored(DetailAST ast) {
-                && (excludeScope == null
-                    || !customScope.isIn(excludeScope)
-                    || !surroundingScope.isIn(excludeScope));
+            try {
+                final AccessModifierOption accessModifier =
+                    CheckUtil.getAccessModifierFromModifiersToken(ast);
+                result = matchAccessModifiers(accessModifier);
+            }
+            catch (IllegalArgumentException ex) {
+                result = false;
+            }
```
Please explain why we are using a try/catch here, this seems hacky.

kkoutsilis replied:
An IllegalArgumentException is thrown in [getAccessModifierFromModifiersTokenDirectly](https://github.com/checkstyle/checkstyle/blob/d4025b26159e5e386eaeb3abd4011bd990b39284/src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java#L341-L345) which is used in `getAccessModifierFromModifiersToken`. As I tried to re-use existing functionality, the exception is caught here and the check is skipped, as this is what should happen to the best of my understanding.

kkoutsilis replied:
I believe this actually is causing the issue with the Enum ([ref](https://github.com/checkstyle/checkstyle/pull/16049#discussion_r1966988478)), looking into it.

nrmancuso replied:
>Is there any concern with this approach from your side?

Absolutely. Exceptions should happen when something exceptional occurs. If CheckUtil#getAccessModifierFromModifiersToken throws an exception when trying to get the modifiers from an enum constant, this is a bug and we should update CheckUtil#getAccessModifierFromModifiersToken.

I presume that CheckUtil#getAccessModifierFromModifiersToken handles other implied modifiers, like on fields in interfaces, right? If so, it should also handle implied modifiers on enum constants.

kkoutsilis replied:
@nrmancuso You are right, thanks for pointing that out.
 To summarize: 
- [CheckUtil#getAccessModifierFromModifiersToken](https://github.com/checkstyle/checkstyle/blob/d4025b26159e5e386eaeb3abd4011bd990b39284/src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java#L316-L330) is trying to `findFirstToken` of token type MODIFIERS
- For ENUM_CONSTANT_DEF, there is no MODIFIERS token type, instead, there is [ANNOTATIONS](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5033-L5065)
- As per [Java Language Specification](https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.9.3) (and also [pointed out by you](https://github.com/checkstyle/checkstyle/pull/16049#discussion_r1966988478)), enum constants are implicitly  declared `public static final`
- For that, `getAccessModifierFromModifiersToken` needs to be updated to handle this use case

Hopefully, I got that right. 
The code has been [updated](https://github.com/checkstyle/checkstyle/compare/d952ec79ee0f7bb07c23ffe1ff4dced69bc3e931..d17f97b70657f159ff1a5f44ac500e0f492b975d). 
Will be waiting for your feedback and thanks again 🙏

nrmancuso commented on src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/javadoc/JavadocVariableCheck.xml:
```diff
@@ -8,18 +8,15 @@
+            <property default-value="public, protected, package, private"
+                      name="accessModifiers"
+                      type="com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption[]">
+               <description>Access modifiers of methods where parameters are
+ checked.</description>
```
Same as above, this doesn't make any sense. Please update everywhere that mentions methods and parameters, this check has nothing to do with these.

nrmancuso commented on src/site/xdoc/checks/javadoc/javadocvariable.xml:
```diff
@@ -78,33 +71,33 @@
 </code></pre></div>
         <p id="Example1-code">
           By default, this setting will report a violation if
-          there is no javadoc for any scope member.
+          there is no javadoc for a member with any access modifier.
```
```
By default, this configuration reports a violation for any field, regardless of its visibility, if it is missing a Javadoc.
```

nrmancuso commented on src/site/xdoc/checks/javadoc/javadocvariable.xml:
```diff
@@ -78,33 +71,33 @@
         <p id="Example2-config">
-          To configure the check for <code>public</code>
-          scope:
+          To configure the check <code>public</code>
+          access modifier:
```
```suggestion
To configure the check to report violations for fields with the <code>public</code> access modifier that are missing Javadocs:
```

nrmancuso replied:
@kkoutsilis missed this one, you should be replying to each and every review item so that we do not miss anything.

kkoutsilis replied:
Sorry for that, I thought I addressed that last time. 
[Addressed](https://github.com/checkstyle/checkstyle/compare/d952ec79ee0f7bb07c23ffe1ff4dced69bc3e931..d17f97b70657f159ff1a5f44ac500e0f492b975d)

nrmancuso commented on src/site/xdoc/checks/javadoc/javadocvariable.xml:
```diff
@@ -122,42 +115,40 @@ public class Example2 {
         <p id="Example3-config">
-          To configure the check for members which are in <code>private</code>, but not in
-          <code>protected</code> scope:
+          To configure the check for members which are in <code>private</code> or
+          <code>package</code> access modifier:
```
```suggestion
To configure the check for fields with the <code>private</code or package-private access modifier:
```

`<code>package</code>` should not be in code tags, this is not valid code.

nrmancuso replied:
Also missed this one

kkoutsilis replied:
[Addressed](https://github.com/checkstyle/checkstyle/compare/d952ec79ee0f7bb07c23ffe1ff4dced69bc3e931..d17f97b70657f159ff1a5f44ac500e0f492b975d)

nrmancuso commented on src/site/xdoc/checks/javadoc/javadocvariable.xml.template:
```diff
@@ -35,7 +35,7 @@
         <p id="Example1-code">
           By default, this setting will report a violation if
-          there is no javadoc for any scope member.
+          there is no javadoc for a member with any access modifier.
```
Please change all mention of `member` or `variable` to `field` in the docs to be consistent.

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableInner.java:
```diff
@@ -1,11 +1,9 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Isn't the default all access modifiers? This doesn't make sense. @romani do we have a ticket somewhere to add validation for this?

romani replied:
Validation is already in place, we might have defect in it. It is possible.

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableInner2.java:
```diff
@@ -16,24 +14,22 @@
     {
-        // Ignore - should be all upper case
+
+        /** Ignore - 1 violation */
```
This javadoc is confusing here. Why does this say "Ignore"?

kkoutsilis replied:
It already exists in the same file a bit lower https://github.com/checkstyle/checkstyle/blob/ad015edbde5a75fd817a49738a529d88dfbc6c67/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableInner2.java#L66 . Don't believe it has a specific reason to have Ignore in it.

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableInterfaceMemberScopeIsPublic.java:
```diff
@@ -15,16 +13,16 @@ public interface InputJavadocVariableInterfaceMemberScopeIsPublic {
         A,
 
-        B; // violation
+        B;
```
The constant above and this one should be a violation; it is equivalent to:
```
public static final Enum B = new Enum();
```

kkoutsilis replied:
This comes up as an `ENUM_CONSTANT_DEF`, and we check for `VARIABLE_DEF`. I guess we still want to check for this case?

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableMethodInnerClass.java:
```diff
@@ -0,0 +1,32 @@
+    private int variablePrivate; // violation, 'Missing a Javadoc comment'
+
+    public void testMethodInnerClass() {
+        class InnerClass {
```
```suggestion
        // this check ignores locally declared classes
        class InnerClass {
```

kkoutsilis replied:
Added

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableMethodInnerClass.java:
```diff
@@ -0,0 +1,32 @@
+            int innerClassVariablePackage;
+            private int innerClassVariablePrivate;
+        }
+        Runnable runnable = new Runnable() {
```
```suggestion
        // this check ignores anonymous inner classes
        Runnable runnable = new Runnable() {
```

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableMethodInnerClass.java:
We need to add both of these examples in the above comments to the documentation to make it clear that we don't check these.

kkoutsilis replied:
Added

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableNoJavadoc4.java:
```diff
@@ -1,32 +1,30 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Incorrect default, again

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableNoJavadoc5.java:
```diff
@@ -1,32 +1,30 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Incorrect default, again

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableNoJavadocNeededInLambda.java:
```diff
@@ -1,19 +1,18 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Incorrect default again

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableOnInnerClassFields.java:
```diff
@@ -1,32 +1,30 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Same as above

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableOnPublicInnerClassFields.java:
```diff
@@ -1,31 +1,29 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Same as above

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariablePublicOnly.java:
```diff
@@ -1,11 +1,9 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Same as above

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableTags.java:
```diff
@@ -1,11 +1,9 @@
 JavadocVariable
-scope = (default)private
-excludeScope = (default)null
+accessModifiers = (default)private
```
Same as above

kkoutsilis replied:
fixed

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableTags.java:
```diff
@@ -308,14 +306,14 @@ public int foo(Object _arg) {
 {
-    CONSTANT_A, // violation
+    CONSTANT_A,
 
     /**
      *
      */
     CONSTANT_B,
 
-    CONSTANT_C // violation
+    CONSTANT_C
```
These should both be violations (CONSTANT_A and CONSTANT_C)

kkoutsilis replied:
fixed

nrmancuso reviewed (changes requested):
First pass, items:

romani commented on .ci/validation.sh:
```diff
@@ -216,12 +216,17 @@ no-error-pmd)
   echo "Checkout target sources ..."
-  checkout_from "https://github.com/pmd/build-tools.git"
+  #  checkout_from "https://github.com/pmd/build-tools.git"
+  checkout_from "https://github.com/kkoutsilis/build-tools.git"
```
Yes, we can not wait. We will be back ones they merge.

nrmancuso replied:
>Yes, we can not wait. We will be back ones they merge.

How will we track this?

romani replied:
Branch will be removed after merge, and CI will fail. I am subscribed to such PR, so eventually I will migrate back.

We can create issue but, still bunch of room to miss. What i described above usually works.

romani commented:
@kkoutsilis , please address review items, lets me know if something is complicated.

romani commented:
Result of test execution should be in Commit.
Test updates all for xdoc
https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=26570&view=logs&j=9c723939-9148-5f6a-3101-ca017cb22518&t=c03cca9e-b908-564b-f216-90279d42b489&l=1963

kkoutsilis commented:
@nrmancuso thanks for your thorough review, appreciate it. I have addressed most, if not all, of your comments. Please have another look when you have some time and let me know if you are happy or if there is anything else to discuss/change.

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocvariable/InputJavadocVariableInner2.java:
```diff
@@ -16,24 +14,23 @@
     interface InnerInterface2
     {
-        // Ignore - should be all upper case
+
+        /** Ignore - should be all upper case. */
```
Please remove this comment, it is nonsense.

kkoutsilis replied:
[Removed](https://github.com/checkstyle/checkstyle/compare/d952ec79ee0f7bb07c23ffe1ff4dced69bc3e931..d17f97b70657f159ff1a5f44ac500e0f492b975d)

nrmancuso reviewed (changes requested):
Items from previous review that were missed:
 - https://github.com/checkstyle/checkstyle/pull/16049#discussion_r1966983412
 - https://github.com/checkstyle/checkstyle/pull/16049#discussion_r1966984545

Item to fix/improve: https://github.com/checkstyle/checkstyle/pull/16049#discussion_r1966978443

New item:

nrmancuso reviewed (approved):
Good to merge after CI is green

@kkoutsilis pitest is failing, please fix this

romani commented:
Yes, hard blocker.
Extra test is required, we did and keep  doing huge  effort to kill all pitest. Let us know if there is problems.

kkoutsilis commented:
@romani @nrmancuso Made a small change to `getAccessModifierFromModifiersToken` to remove the pitest mutation, please have another look and let me know if this is fine. Thank you

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtil.java:
```diff
@@ -314,9 +314,14 @@ public static boolean isReceiverParameter(DetailAST parameterDefAst) {
-                getAccessModifierFromModifiersTokenDirectly(modsToken);
+        AccessModifierOption accessModifier;
+        if (ast.getType() == TokenTypes.ENUM_CONSTANT_DEF) {
+            accessModifier = AccessModifierOption.PUBLIC;
```
Why we think it is always public?
Please share a bit more details, thanks in advance.

nrmancuso replied:
Enum constants are always public, static, and final. This cannot be changed/overridden in any way.

romani replied:
Yes, I read it as enum itself.

romani reviewed (changes requested):
Question

romani reviewed (approved):
Let's merge this huge update.

romani commented:
@kkoutsilis , please send revert on one your fork usage. Migration is done on their side.
